ICE: `entered unreachable code: no sub-expr expected for Pat`

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' panicked at compiler/rustc_hir_typeck/src/expr.rs:430:17: 'internal error: entered unreachable code: no sub-expr expected for Pat(Pat { hir_id: HirId(DefId(0:3 ~ a[4e31]::test2).15), kind: Range(Some(Expr { hir_id: HirId(DefId(0:3 ~ a[4e31]::test2).16), kind: Lit(Spanned { node: Int(Pu128(0), Unsuffixed), span: /tmp/im/a.rs:7:9: 7:10 (#0) }), span: /tmp/im/a.rs:7:9: 7:10 (#0) }), Some(Expr { hir_id: HirId(DefId(0:3 ~ a[4e31]::test2).17), kind: Path(Resolved(None, Path { span: /tmp/im/a.rs:7:15: 7:16 (#0), res: Local(HirId(DefId(0:3 ~ a[4e31]::test2).5)), segments: [PathSegment { ident: x#0, hir_id: HirId(DefId(0:3 ~ a[4e31]::test2).18), res: Local(HirId(DefId(0:3 ~ a[4e31]::test2).5)), args: None, infer_args: true }] })), span: /tmp/im/a.rs:7:15: 7:16 (#0) }), Included), span: /tmp/im/a.rs:7:9: 7:16 (#0), default_binding_modes: true })'', 'thread 'rustc' panicked at compiler/rustc_hir_typeck/src/expr.rs:430:17: 'internal error: entered unreachable code: no sub-expr expected for Pat(Pat { hir_id: HirId(DefId(0:3 ~ a[4e31]::test2).15), kind: Range(Some(Expr { hir_id: HirId(DefId(0:3 ~ a[4e31]::test2).16), kind: Lit(Spanned { node: Int(Pu128(0), Unsuffixed), span: /tmp/im/a.rs:7:9: 7:10 (#0) }), span: /tmp/im/a.rs:7:9: 7:10 (#0) }), Some(Expr { hir_id: HirId(DefId(0:3 ~ a[4e31]::test2).17), kind: Path(Resolved(None, Path { span: /tmp/im/a.rs:7:15: 7:16 (#0), res: Local(HirId(DefId(0:3 ~ a[4e31]::test2).5)), segments: [PathSegment { ident: x#0, hir_id: HirId(DefId(0:3 ~ a[4e31]::test2).18), res: Local(HirId(DefId(0:3 ~ a[4e31]::test2).5)), args: None, infer_args: true }] })), span: /tmp/im/a.rs:7:15: 7:16 (#0) }), Included), span: /tmp/im/a.rs:7:9: 7:16 (#0), default_binding_modes: true })''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
fn test2() {
    let x: !;
    let c2 = SingleVariant::Points(0)
        | match 1 {
            0..=x => {}
        };
}

````

original:
````rust
#![feature(never_type)]

fn test2() {
    let x: !;
    let c2 = SingleVariant::Points(0)
        | match 1 {
        0 ..= x => {}
        //~^ ERROR: runtime values cannot be referenced in patterns
        0 ..= FOO => {}
        //~^ ERROR: statics cannot be referenced in patterns
    };
}

enum B { B }

fn main() {}

````

Version information
````
rustc 1.85.0-nightly (c94848c04 2024-12-05)
binary: rustc
commit-hash: c94848c046d29f9a80c09aae758e27e418a289f2
commit-date: 2024-12-05
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.5

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/c94848c046d29f9a80c09aae758e27e418a289f2/compiler/rustc_hir_typeck/src/expr.rs#L424-L436

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: the `!` type is experimental
 --> /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:2:12
  |
2 |     let x: !;
  |            ^
  |
  = note: see issue #35121 <https://github.com/rust-lang/rust/issues/35121> for more information
  = help: add `#![feature(never_type)]` to the crate attributes to enable
  = note: this compiler was built on 2024-12-05; consider upgrading it if it is out of date

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:7:2
  |
7 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs`

error[E0433]: failed to resolve: use of undeclared type `SingleVariant`
 --> /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:3:14
  |
3 |     let c2 = SingleVariant::Points(0)
  |              ^^^^^^^^^^^^^ use of undeclared type `SingleVariant`

thread 'rustc' panicked at compiler/rustc_hir_typeck/src/expr.rs:430:17:
internal error: entered unreachable code: no sub-expr expected for Pat(Pat { hir_id: HirId(DefId(0:3 ~ mvce[ef52]::test2).15), kind: Range(Some(Expr { hir_id: HirId(DefId(0:3 ~ mvce[ef52]::test2).16), kind: Lit(Spanned { node: Int(Pu128(0), Unsuffixed), span: /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:5:13: 5:14 (#0) }), span: /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:5:13: 5:14 (#0) }), Some(Expr { hir_id: HirId(DefId(0:3 ~ mvce[ef52]::test2).17), kind: Path(Resolved(None, Path { span: /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:5:17: 5:18 (#0), res: Local(HirId(DefId(0:3 ~ mvce[ef52]::test2).5)), segments: [PathSegment { ident: x#0, hir_id: HirId(DefId(0:3 ~ mvce[ef52]::test2).18), res: Local(HirId(DefId(0:3 ~ mvce[ef52]::test2).5)), args: None, infer_args: true }] })), span: /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:5:17: 5:18 (#0) }), Included), span: /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:5:13: 5:18 (#0), default_binding_modes: true })
stack backtrace:
   0:     0x723d3353f0fa - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h6672795076376ec8
   1:     0x723d33c13e26 - core::fmt::write::h792a60ae9bb2dbda
   2:     0x723d34c08391 - std::io::Write::write_fmt::hfc5b1a97b7b0eb31
   3:     0x723d3353ef52 - std::sys::backtrace::BacktraceLock::print::h2c102308db74512c
   4:     0x723d3354146a - std::panicking::default_hook::{{closure}}::h63a255606d7053aa
   5:     0x723d335412b3 - std::panicking::default_hook::h114da533e3893438
   6:     0x723d326b9068 - std[339c42a0e0e1cbc8]::panicking::update_hook::<alloc[4b645d2b808d60e5]::boxed::Box<rustc_driver_impl[a450451a1b077cae]::install_ice_hook::{closure#0}>>::{closure#0}
   7:     0x723d33541c28 - std::panicking::rust_panic_with_hook::hadbdcb738271374d
   8:     0x723d3354191a - std::panicking::begin_panic_handler::{{closure}}::hf015953bfb7b98c8
   9:     0x723d3353f5a9 - std::sys::backtrace::__rust_end_short_backtrace::h8eb8f61d86c09a66
  10:     0x723d335415dd - rust_begin_unwind
  11:     0x723d301afd10 - core::panicking::panic_fmt::h8558f71ff16f773d
  12:     0x723d33d70d6b - <rustc_hir_typeck[17bac021c3c291fe]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  13:     0x723d33df968c - <rustc_hir_typeck[17bac021c3c291fe]::fn_ctxt::FnCtxt>::check_pat
  14:     0x723d34d3cbc4 - <rustc_hir_typeck[17bac021c3c291fe]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
  15:     0x723d33d6dea6 - <rustc_hir_typeck[17bac021c3c291fe]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  16:     0x723d33d8776f - <rustc_hir_typeck[17bac021c3c291fe]::fn_ctxt::FnCtxt>::check_overloaded_binop
  17:     0x723d34d3cc95 - <rustc_hir_typeck[17bac021c3c291fe]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
  18:     0x723d33d6dea6 - <rustc_hir_typeck[17bac021c3c291fe]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  19:     0x723d33d68eb9 - <rustc_hir_typeck[17bac021c3c291fe]::fn_ctxt::FnCtxt>::check_decl
  20:     0x723d33d6b6a1 - <rustc_hir_typeck[17bac021c3c291fe]::fn_ctxt::FnCtxt>::check_expr_block
  21:     0x723d34d3b4e9 - <rustc_hir_typeck[17bac021c3c291fe]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args::{closure#0}
  22:     0x723d33d6dea6 - <rustc_hir_typeck[17bac021c3c291fe]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  23:     0x723d340134e6 - rustc_hir_typeck[17bac021c3c291fe]::check::check_fn
  24:     0x723d3401b441 - rustc_hir_typeck[17bac021c3c291fe]::typeck
  25:     0x723d3401a05d - rustc_query_impl[3a63db0db156c2ea]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3a63db0db156c2ea]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[aab370552e9df2de]::query::erase::Erased<[u8; 8usize]>>
  26:     0x723d34026f11 - rustc_query_system[924318b55d1b9da5]::query::plumbing::try_execute_query::<rustc_query_impl[3a63db0db156c2ea]::DynamicConfig<rustc_data_structures[904645e0e499155f]::vec_cache::VecCache<rustc_span[cf656b7144c9c5c5]::def_id::LocalDefId, rustc_middle[aab370552e9df2de]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[924318b55d1b9da5]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[3a63db0db156c2ea]::plumbing::QueryCtxt, false>
  27:     0x723d340261cd - rustc_query_impl[3a63db0db156c2ea]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  28:     0x723d34025e7f - <rustc_middle[aab370552e9df2de]::hir::map::Map>::par_body_owners::<rustc_hir_analysis[c0842703e669cd]::check_crate::{closure#4}>::{closure#0}
  29:     0x723d34023e92 - rustc_hir_analysis[c0842703e669cd]::check_crate
  30:     0x723d3402e9c4 - rustc_interface[b7068c2f67f9e538]::passes::run_required_analyses
  31:     0x723d34be5f9e - rustc_interface[b7068c2f67f9e538]::passes::analysis
  32:     0x723d34be5f6f - rustc_query_impl[3a63db0db156c2ea]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3a63db0db156c2ea]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[aab370552e9df2de]::query::erase::Erased<[u8; 1usize]>>
  33:     0x723d34bbdd7a - rustc_query_system[924318b55d1b9da5]::query::plumbing::try_execute_query::<rustc_query_impl[3a63db0db156c2ea]::DynamicConfig<rustc_query_system[924318b55d1b9da5]::query::caches::SingleCache<rustc_middle[aab370552e9df2de]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[3a63db0db156c2ea]::plumbing::QueryCtxt, false>
  34:     0x723d34bbda4e - rustc_query_impl[3a63db0db156c2ea]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  35:     0x723d34c9142b - rustc_interface[b7068c2f67f9e538]::interface::run_compiler::<core[a08a8da589f49b02]::result::Result<(), rustc_span[cf656b7144c9c5c5]::ErrorGuaranteed>, rustc_driver_impl[a450451a1b077cae]::run_compiler::{closure#0}>::{closure#1}
  36:     0x723d34adc387 - std[339c42a0e0e1cbc8]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[b7068c2f67f9e538]::util::run_in_thread_with_globals<rustc_interface[b7068c2f67f9e538]::util::run_in_thread_pool_with_globals<rustc_interface[b7068c2f67f9e538]::interface::run_compiler<core[a08a8da589f49b02]::result::Result<(), rustc_span[cf656b7144c9c5c5]::ErrorGuaranteed>, rustc_driver_impl[a450451a1b077cae]::run_compiler::{closure#0}>::{closure#1}, core[a08a8da589f49b02]::result::Result<(), rustc_span[cf656b7144c9c5c5]::ErrorGuaranteed>>::{closure#0}, core[a08a8da589f49b02]::result::Result<(), rustc_span[cf656b7144c9c5c5]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a08a8da589f49b02]::result::Result<(), rustc_span[cf656b7144c9c5c5]::ErrorGuaranteed>>
  37:     0x723d34adc022 - <<std[339c42a0e0e1cbc8]::thread::Builder>::spawn_unchecked_<rustc_interface[b7068c2f67f9e538]::util::run_in_thread_with_globals<rustc_interface[b7068c2f67f9e538]::util::run_in_thread_pool_with_globals<rustc_interface[b7068c2f67f9e538]::interface::run_compiler<core[a08a8da589f49b02]::result::Result<(), rustc_span[cf656b7144c9c5c5]::ErrorGuaranteed>, rustc_driver_impl[a450451a1b077cae]::run_compiler::{closure#0}>::{closure#1}, core[a08a8da589f49b02]::result::Result<(), rustc_span[cf656b7144c9c5c5]::ErrorGuaranteed>>::{closure#0}, core[a08a8da589f49b02]::result::Result<(), rustc_span[cf656b7144c9c5c5]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a08a8da589f49b02]::result::Result<(), rustc_span[cf656b7144c9c5c5]::ErrorGuaranteed>>::{closure#1} as core[a08a8da589f49b02]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  38:     0x723d34adb76b - std::sys::pal::unix::thread::Thread::new::thread_start::hc40c233155f425ee
  39:     0x723d2eea339d - <unknown>
  40:     0x723d2ef2849c - <unknown>
  41:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.85.0-nightly (c94848c04 2024-12-05) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `test2`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0433, E0601, E0658.
For more information about an error, try `rustc --explain E0433`.

```

</p>
</details>

<!--
query stack:
internal error: entered unreachable code: no sub-expr expected for Pat(Pat { hir_id: HirId(DefId(0:3 ~ mvce[ef52]::test2).15), kind: Range(Some(Expr { hir_id: HirId(DefId(0:3 ~ mvce[ef52]::test2).16), kind: Lit(Spanned { node: Int(Pu128(0), Unsuffixed), span: /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:5:13: 5:14 (#0) }), span: /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:5:13: 5:14 (#0) }), Some(Expr { hir_id: HirId(DefId(0:3 ~ mvce[ef52]::test2).17), kind: Path(Resolved(None, Path { span: /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:5:17: 5:18 (#0), res: Local(HirId(DefId(0:3 ~ mvce[ef52]::test2).5)), segments: [PathSegment { ident: x#0, hir_id: HirId(DefId(0:3 ~ mvce[ef52]::test2).18), res: Local(HirId(DefId(0:3 ~ mvce[ef52]::test2).5)), args: None, infer_args: true }] })), span: /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:5:17: 5:18 (#0) }), Included), span: /tmp/icemaker_global_tempdir.j7zmU1Vhf0Zs/rustc_testrunner_tmpdir_reporting.1W4prXpGdnHk/mvce.rs:5:13: 5:18 (#0), default_binding_modes: true })
#0 [typeck] type-checking `test2`
#1 [analysis] running analysis passes on this crate

-->
@rustbot label +F-never_type

--- a/compiler/rustc_hir_typeck/src/expr.rs
+++ b/compiler/rustc_hir_typeck/src/expr.rs
@@ -402,6 +402,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             })
             | hir::Node::ImplItem(hir::ImplItem { kind: hir::ImplItemKind::Const(..), .. }) => true,
 
+            hir::Node::Pat(_) => {
+                self.dcx().span_delayed_bug(expr.span, "place expr not allowed in pattern");
+                true
+            }
+
             // These nodes do not have direct sub-exprs.
             hir::Node::Param(_)
             | hir::Node::Item(_)
@@ -414,7 +419,6 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             | hir::Node::Ty(_)
             | hir::Node::AssocItemConstraint(_)
             | hir::Node::TraitRef(_)
-            | hir::Node::Pat(_)
             | hir::Node::PatField(_)
             | hir::Node::LetStmt(_)
             | hir::Node::Synthetic

--- a/tests/ui/never_type/never-in-range-pat.rs
+++ b/tests/ui/never_type/never-in-range-pat.rs
@@ -1,0 +1,16 @@
+// Regression test for <https://github.com/rust-lang/rust/issues/133947>.
+
+// Make sure we don't ICE when there's `!` in a range pattern.
+//
+// This shouldn't be allowed anyways, but we only deny it during MIR
+// building, so make sure we handle it semi-gracefully during typeck.
+
+#![feature(never_type)]
+
+fn main() {
+    let x: !;
+    match 1 {
+        0..x => {}
+        //~^ ERROR only `char` and numeric types are allowed in range patterns
+    }
+}

--- a/tests/ui/never_type/never-in-range-pat.stderr
+++ b/tests/ui/never_type/never-in-range-pat.stderr
@@ -1,0 +1,11 @@
+error[E0029]: only `char` and numeric types are allowed in range patterns
+  --> $DIR/never-in-range-pat.rs:13:12
+   |
+LL |         0..x => {}
+   |         -  ^ this is of type `!` but it should be `char` or numeric
+   |         |
+   |         this is of type `{integer}`
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0029`.

